### ISSUE TITLE
[Solved] Provide list of all connection parameters allowed in the "parameters" string

### ISSUE THREAD

**Issue by koalabi** [Dzordzu/rust-apache-age#9]
For evaluation purposes, I'm using the official docker image where the default database is postgresDB but I cannot access this database using the (adapted) example code. If just adapt the values, I get the error "database postgresUser does not exist" (due to PostgreSQL's habit, I think, to assume that the default database has the name of the user).

```
thread 'main' panicked at src/main.rs:19:6:
called `Result::unwrap()` on an `Err` value: Error { kind: Db, cause: Some(DbError { severity: "FATAL", parsed_severity: Some(Fatal), code: SqlState(E3D000), message: "database \"postgresUser\" does not exist", detail: None, hint: None, position: None, where_: None, schema: None, table: None, column: None, datatype: None, constraint: None, file: Some("postinit.c"), line: Some(855), routine: Some("InitPostgres") }) }

```

I have also tried to add to the string "database=postgresDB" or even "schema=postgresDB" but I get errors in both cases:

```
thread 'main' panicked at src/main.rs:19:6:
called `Result::unwrap()` on an `Err` value: Error { kind: ConfigParse, cause: Some(UnknownOption("database")) }
```
 and

```
thread 'main' panicked at src/main.rs:19:6:
called `Result::unwrap()` on an `Err` value: Error { kind: ConfigParse, cause: Some(UnknownOption("schema")) }
```
Code is typically:
```rust
    // Create client
    let mut client = Client::connect_age(
        "host=localhost user=postgresUser password=postgresPW port=5455",
        NoTls,
    )
    .unwrap();
```
with, of course, variations in function of the attempt (database, schema, ...)

The parameter values are correct because the database is accessible with these parameters through AGE-Viewer (that passes the following parameters from browser to age-viewer's backend (that in turn calls AGE):

```yml
{
    "host": "localhost",
    "port": 5455,
    "database": "postgresDB",
    "user": "postgresUser",
    "password": "postgresPW"
}
```
I have looked at the source code but found absolutely nothing. So, I'm stuck :-(.

This is at best a documentation error (if parameter names are verbatim the ones of AGE, OK, they don't need to be documented in extenso in the driver documentation but at least a reference to the full parameter list should be provided (I have searched on the AGE site but found nothing).

It can also be a bug ... but I haven't found where ...

Any help is welcome. Thanks in advance.

## Solution

* https://github.com/Dzordzu/rust-apache-age/issues/9#issuecomment-1928007159
* https://github.com/Dzordzu/rust-apache-age/issues/9#issuecomment-1929882460

**Comment by Dzordzu** [comment]
Hi. Can you provide more details for "official docker image". I would like to know the name of the image, tag, version of the postgresql, version of the AGE and (optionally) checksum of the image.

**Comment by koalabi** [comment]
> Hi. Can you provide more details for "official docker image". I would like to know the name of the image, tag, version of the postgresql, version of the AGE and (optionally) checksum of the image. Also, either the docker-compose file, or the docker cmd that starts your container would be quite handful.

The Docker image for AGE is this one: https://hub.docker.com/r/apache/age:PG11_latest
The image for AGE-Viewer is this one: https://hub.docker.com/r/nicered/age-viewer:latest

Both images were instantiated as containers with almost all default values except for ports and credentials.

**Comment by Dzordzu** [comment]
> https://hub.docker.com/r/apache/age:PG11_latest

404. Dead link.

> https://hub.docker.com/r/nicered/age-viewer:latest

404. Dead link.

**Comment by koalabi** [comment]
It is just the tail (after the ':'), i.e. the version tag that poses problem; I had added them to specify the version.

These links work
https://hub.docker.com/r/apache/age
https://hub.docker.com/r/nicered/age-viewer

I don't use a docker-compose in my case; just 2 separate containers but, as stated in my initial mail, there is no problem with any container. Via the containers, I can use AGE-Viewer and the database correctly. It is the CLI example from the driver's source tree that causes panic ...

**Comment by koalabi** [comment]
I just need to know what string I should use as "parameters" with your driver if the database is not just "postgres" ... If I look in the browser's debug console, I see the parameters that work but they don't work if I use them in the CLI program.

The program panics in the code fragment I show here above ...

**Comment by Dzordzu** [comment]
## Some significant notes before answer

1. When somebody asks you for information, just provide it. Otherwise you are just wasting time of yours and others. This sucks.
2.  If you post links, make sure they work. People don't want to figure out what you want. I'm not inside your mind to guess what you've meant. 
3. According to your accusations

> It is the CLI example from the driver's source tree that causes panic

Nope. It's you.  The example has been working for me like a charm.

4. Furthermore,

> I just need to know what string I should use as "parameters" 

I recommend [reading the docs](https://docs.rs/apache_age/latest/apache_age/), where the first line actually answers your question:

> Connector to the [apache age](https://age.apache.org/) database is based on the postgres and tokio-postgres crates.

 [The first link in the google (rust cookbook)](https://rust-lang-nursery.github.io/rust-cookbook/database/postgres.html) solves your issue
 
 ## "Why so negative attitude"
 
 When asking for help / reporting a bug, one should provide *compact*, *clear* and *detailed* information about environment and issue background. I'm full of work, but I love helping people. Sadly, my time is limited. So are my nerves.
 
 ## The solution
 
 <details open>
<summary>docker-compose.yml (for reference)</summary>
<br />

```yml
---
services:
  age:
    image: apache/age:PG11_latest
    environment:
      POSTGRES_PASSWORD: passwd
      POSTGRES_DB: myDB123
      POSTGRES_USER: pgUser
    ports:
      - "9081:5432"
  viewer:
    image: bitnine/agviewer:latest
    ports:
      - "9080:3001"
```

</details>
 
 ```rust
     let mut client = Client::connect_age(
        "postgresql://pgUser:passwd@localhost:9081/myDB123",
        NoTls,
    )
    .unwrap();
 ```

**Comment by koalabi** [comment]
OK. Let's stop here. You are not reading, you are not listening. You are asking for information that is totally irrelevant.

I will pass my way and try something else.

I have probably opened about 30 issues (usually one per repo, not more: I know how it works. I won't provide you with information I don't have (because it is not related to the root cause of my issue and I just don't have it at hand).

**Comment by koalabi** [comment]
No way to discuss.

**Comment by koalabi** [comment]
Issue can be closed.

**Comment by Dzordzu** [comment]
Again. I've reproduced you "issue" and gave you a solution. I've also updated the README to mention clear relationship with the `postgres` and `tokio-postgres` crates

## Solution (once again)

```rust
    let mut client = Client::connect_age(
       "postgresql://pgUser:passwd@localhost:9081/myDB123",
       NoTls,
   )
   .unwrap();
   ```
   
 ## More reading
  
* [postgres::Client docs](https://docs.rs/postgres/latest/postgres/struct.Client.html)
* [postgres::config::Config docs](https://docs.rs/postgres/latest/postgres/config/struct.Config.html)

**Comment by koalabi** [comment]
Great! Thank you. 

I can agree with you (hope so, at least). Let's not engage in a flame war ;-) but the problem was where I suspected it was: in the connection string. (and there was no need to question the setup, that worked) 

f you read my initial report, I mentioned that it might be a documentation issue, what it was in the end. All in all, I was not even aware that it was possible to use so a lazy connection string with PostgreSQL. The one you show in the solution looks so more familiar and also stricter. I hadn't tried it as I suspected there was some kind of translation along the path (that I, indeed, didn't find in your code). Your fix (in the README.md) seems very relevant to me. If I may, I would even recommend that the examples use the "normal/strict" form of connection string. It will probably avoid other (connection-related) issues.

Thanks again.
